File: app/src/main/java/rocks/codesherpas/nostatements/java/CancelReservation.java
package rocks.codesherpas.nostatements.java;

import rocks.codesherpas.nostatements.kotlin.GuestId;

import java.time.DayOfWeek;
import java.time.ZonedDateTime;
import java.util.logging.Logger;

import static java.time.DayOfWeek.SUNDAY;

public class CancelReservation {
    private final ReservationRepository reservationRepository;
    private final ReservationSearcher reservationSearcher;
    private final EventPublisher eventPublisher;

    private static final Logger LOGGER = Logger.getLogger(CancelReservation.class.getName());

    public CancelReservation(ReservationRepository reservationRepository, ReservationSearcher reservationSearcher, EventPublisher eventPublisher) {
        this.reservationRepository = reservationRepository;
        this.reservationSearcher = reservationSearcher;
        this.eventPublisher = eventPublisher;
    }

    public void invokeWith(ReservationId reservationId, GuestId guestId) {
        Reservation reservation = reservationSearcher.lastBy(guestId);

        if (reservation != null) {
            if (reservation.isFinished()) {
                Reservation reservation1 = reservationRepository.findReservationById(reservationId);
                reservation1.cancel();

                eventPublisher.publish(reservation.pullEvents());
                LOGGER.info(formatLogMessage(reservation1));
            }
        } else {
            cancelWithPenalty(reservationId, guestId);
        }
    }

    private void cancelWithPenalty(ReservationId reservationId, GuestId guestId) {
        // apply some cancellation mechanism
    }

    private String formatLogMessage(Reservation reservation1) {
        return "Reservation with id: %s was cancelled.".formatted(reservation1.getReservationId());
    }

    private DayOfWeek today() {
        return ZonedDateTime.now().getDayOfWeek();
    }
}

File: app/src/main/kotlin/rocks/codesherpas/nostatements/kotlin/CancelReservation.kt
package rocks.codesherpas.nostatements.kotlin

import rocks.codesherpas.nostatements.java.EventPublisher
import rocks.codesherpas.nostatements.java.ReservationRepository
import rocks.codesherpas.nostatements.java.ReservationSearcher
import java.time.DayOfWeek
import java.time.DayOfWeek.SUNDAY
import java.time.ZonedDateTime
import java.util.logging.Logger

class CancelReservation(
    private val reservationRepository: ReservationRepository,
    private val reservationSearcher: ReservationSearcher,
    private val eventPublisher: EventPublisher
) {
    companion object {
        private val LOGGER = Logger.getLogger(CancelReservation::class.java.name)
    }

    fun invokeWith(reservationId: ReservationId, guestId: GuestId) {
        reservationSearcher.lastBy(guestId)
            ?.let { lastReservation ->
            if (today() != SUNDAY && lastReservation.isFinished) {
                reservationRepository.findReservationById(reservationId)
                    ?.apply { this.cancel() }
                    ?.also { eventPublisher.publish(it.pullEvents()) }
                    ?.also {
                        LOGGER.info("Reservation with id: ${it.reservationId} was cancelled.")
                    } ?: throw InvalidCancelOperationException()
            }
        } ?: cancelWithPenalty(reservationId, guestId)
    }

    private fun cancelWithPenalty(reservationId: ReservationId, guestId: GuestId) {
        // apply some cancellation mechanism
    }

    private fun today(): DayOfWeek = ZonedDateTime.now().dayOfWeek
}
